source/common/network/win32_socket_handle_impl.cc

```diff
@@ -71,10 +79,43 @@ Api::IoCallUint64Result Win32SocketHandleImpl::recvmmsg(RawSliceArrays& slices,
 }
 
 Api::IoCallUint64Result Win32SocketHandleImpl::recv(void* buffer, size_t length, int flags) {
+  if (flags & MSG_PEEK) {
+    return peek(buffer, length);
+  }
 
-  Api::IoCallUint64Result result = IoSocketHandleImpl::recv(buffer, length, flags);
-  reEnableEventBasedOnIOResult(result, Event::FileReadyType::Read);
-  return result;
+  if (peek_buffer_.length() == 0) {
+    Api::IoCallUint64Result result = IoSocketHandleImpl::recv(buffer, length, flags);
+    reEnableEventBasedOnIOResult(result, Event::FileReadyType::Read);
+    return result;
+  } else {
+    return readFromPeekBuffer(buffer, length);
+  }
+}
+
+Api::IoCallUint64Result Win32SocketHandleImpl::peek(void* buffer, size_t length) {
+
+  // if the `peek_buffer_` has already enough data we can skip reading
+  // from the wire.
```

Comment:
I'd be inclined to comment the work directly below.
"If there's not enough data in the peek buffer, try reading more"